include/nighthawk/adaptive_load/step_controller.h

```diff
@@ -0,0 +1,86 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+#include "envoy/config/core/v3/base.pb.h"
+#include "envoy/config/typed_config.h"
+
+#include "external/envoy/source/common/protobuf/protobuf.h"
+
+#include "api/adaptive_load/benchmark_result.pb.h"
+
+namespace Nighthawk {
+namespace AdaptiveLoad {
+
+/**
+ * An interface for StepControllers that compute load adjustments and check for convergence.
+ *
+ * See source/adaptive_load/step_controller_impl.h for example plugins.
+ */
+class StepController {
+public:
+  virtual ~StepController() = default;
+  /**
+   * Returns the current CommandLineOptions load specification that the StepController recommends.
+   *
+   * @return CommandLineOptions The final product after applying all computed load variables via
+   * InputVariableSetter plugins to the stored CommandLineOptions template.
+   */
+  virtual nighthawk::client::CommandLineOptions GetCurrentCommandLineOptions() const PURE;
+  /**
+   * Reports if the search for the optimal load has converged, based on the StepController's
+   * internal state variables.
+   *
+   * @return bool Whether the load has converged.
+   */
+  virtual bool IsConverged() const PURE;
+  /** Reports if the algorithm has determined it can never succeed as configured, e.g. because
+   * metrics were outside thresholds at input values throughout the configured search range.
+   *
+   * @param doom_reason Pointer to a string to write the reason for being doomed. If returning true,
+   * an explanation of why success is impossible is written here; otherwise this string is not
+   * touched.
+   *
+   * @return bool true if the controller has determined convergence is impossible.
+   */
+  virtual bool IsDoomed(std::string* doom_reason) const PURE;
```

Comment:
if doom_reason can be categorized, maybe it is better to have an enum instead of string?

Reply:
In general I expect to need a dynamically generated string to describe the doom reason from a more complicated step controller. So I think we would need either a string or enum+string (like RPC status) to be flexible enough.

Currently with ExponentialSearchStepController there's only one known way to be doomed (threshold already exceeded with initial input value).

We might end up wishing we had a doom code enum later. But it's also some extra work to add it now, and we might not end up needing it.

There's a way to add support for an enum later in a backward compatible way, by adding an extra method:

```cpp
virtual bool IsDoomed(DoomStatus* doom_status) const {
  DoomStatus status;
  status.set_code(UNSPECIFIED_DOOM_CODE);
  std::string doom_reason;
  if (IsDoomed(&doom_reason)) {
    status.set_message(doom_reason);
  }
  return status;
}
```
Which new implementations would override instead of IsDoomed(std::string*). The main code would only call the new IsDoomed.

Do you think we can safely proceed with just the string for now?

Reply:
Thanks Eric for the explanation. Yes, I agree we can proceed with the string.